multicluster/controllers/multicluster/leader_clusterset_controller.go

```diff
@@ -0,0 +1,158 @@
+/*
+Copyright 2021 Antrea Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package multicluster
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	multiclusterv1alpha1 "antrea.io/antrea/multicluster/apis/multicluster/v1alpha1"
+	"antrea.io/antrea/multicluster/controllers/multicluster/common"
+)
+
+// LeaderClusterSetReconciler reconciles a ClusterSet object in the leader cluster deployment.
+// There will be one MC Controller running in each Namespace of the leader for multiple ClusterSet support.
+// So each MC Controller will only be handling a single ClusterSet in the given Namespace.
+type LeaderClusterSetReconciler struct {
+	client.Client
+	mutex  sync.Mutex
+	Scheme *runtime.Scheme
+
+	clusterSetConfig *multiclusterv1alpha1.ClusterSet
+	clusterSetID     common.ClusterSetID
+	clusterID        common.ClusterID
+}
+
+//+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=clustersets,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=clustersets/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=clustersets/finalizers,verbs=update
+
+// Reconcile ClusterSet changes
+func (r *LeaderClusterSetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	clusterSet := &multiclusterv1alpha1.ClusterSet{}
+	err := r.Get(ctx, req.NamespacedName, clusterSet)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+		// if errors.IsNotFound(err)
+		klog.InfoS("Received ClusterSet delete", "config", klog.KObj(clusterSet))
+		r.clusterSetConfig = nil
+		r.clusterID = common.InvalidClusterID
+		r.clusterSetID = common.InvalidClusterSetID
+
+		return ctrl.Result{}, nil
+	}
+
+	klog.InfoS("Received ClusterSet add/update", "config", klog.KObj(clusterSet))
+
+	// Handle create or update
+	// If create, make sure the local ClusterClaim is part of the leader config
```

Comment:
here `validateLocalClusterClaim` will be called not matter it's create or update event, so I am wondering is this your purpose or not? if yes, maybe you need to update the comment: s/If create/If create or update/

Reply:
Could you add unit test codes for member and leader clusterset controller? thanks.

Reply:
Okay added conditions to avoid checking for update.
I will push that change to the new PR to avoid having to rebase a lot.
Looking at testing the clusterset controllers now.